### DISCUSSION TITLE
STM32: necessity to turn off peripherals during stop mode for power saving

### DISCUSSION THREAD

**Discussion by ycsin** [zephyrproject-rtos/zephyr#41639] [General] [answered]
We are working on a battery-operated application where the current consumption (especially during STOP mode) determines the success of the project. We are currently using Joulescope for our measurements.

Yesterday my colleague who tested the `AUTO_OFF` mode of the ADC reported no measurable difference in the current consumption, even though it "can cause a dramatic reduction in the power consumption of applications"
<img width="800" alt="image" src="https://user-images.githubusercontent.com/1563974/148505386-74a200d8-0aab-437a-a19f-8f5784c5d81a.png">

He obtained the following measurements:

### Test 1: ADC is enabled, enabling/disabling `AUTO_OFF` configuration during ADC init.

| `CONFIG_PM` | ADC `AUTO_OFF` | Current(mA) |
|--------------------|-------------------------------------|-------------|
| y                  | y                                   | 0.25        |
| y                  | n                                   | 0.25        |
| n                  | y                                   | 6.65        |
| n                  | n                                   | 6.96        |

### Test 2: Disable the ADC

| `CONFIG_PM` | `LL_ADC_Disable` | Current(mA) |
|--------------------|------------|-------------|
| y                  | y          | 0.25        |
| y                  | n          | 0.25        |
| n                  | y          | 6.65        |
| n                  | n          | 6.96        |

The tests showed that both `LL_ADC_Disable` & `AUTO_OFF` have no effect on the current consumption if the MCU enters STOP mode, but can have a 0.3mA reduction when the CPU is in RUN mode and the ADC is not converting.

He used the following code to perform the test, configuring `AUTO_OFF` / Disable the ADC at the end of ADC init function:
<img width="600" alt="image" src="https://user-images.githubusercontent.com/1563974/148508333-935d65b3-3082-4789-af44-681e3420a127.png">

Digging further, I found this in the PWR chapter of the G0B1RE reference manual
<img width="800" alt="image" src="https://user-images.githubusercontent.com/1563974/148505506-01d8060a-8c6d-4ba4-a040-8855abf3db91.png">
<img width="800" alt="image" src="https://user-images.githubusercontent.com/1563974/148505558-fceaeb66-1366-4fe0-a77b-2fab1750a941.png">
<img width="800" alt="image" src="https://user-images.githubusercontent.com/1563974/148505604-aad2a270-4c46-41bd-b6ce-12f6cff96e39.png">

I'm wondering if the ADC is disabled by hardware when the MCU entered STOP mode? If it does, that explains why there's no difference in current consumption when `AUTO_OFF` is enabled or not.

And SPIs are same as the ADC (not available), if so it explains why I cant measure the difference in my previous PR using a multimeter #36681, and this PR is not necessary since ADC will be turned off automatically by hardware, same goes for #32544

But I'm not sure, and would love some comments.

**Comment by ycsin** [comment]
@erwango @ABOSTM @FRASTM 
Any idea?

**Comment by erwango** [comment, marked as answer]
> I'm wondering if the ADC is disabled by hardware when the MCU entered STOP mode? 

@ycsin This is the case indeed
In stop modes, peripheral clock gating is used to disable peripheral clocks. So most peripherals won't consume power in this state, unless specific configuration is done. Note that, in some configurations, ADC can still consume power in stop mode, but that is not the default case.
For more information, cf for instance RM0444, $4.3.6 "Stop 0 Mode".
